### DISCUSSION TITLE
How to traverse an array of IDisposable? Array as Traversable and IDisposable as Applicative bracket?

### DISCUSSION THREAD

**Discussion by ingun37** [louthy/language-ext#1310] [Q&A] [answered]
Hello. In Haskell, I can traverse (or sequence) when a [bracket](https://wiki.haskell.org/Bracket_pattern) and an array of its input is given. How can I do the same with IDisposable objects in language-ext?

**Comment by louthy** [comment, marked as answer]
For the exact example with the IO monad:

```c#
public static IO<C> bracket<A, B, C>(
    IO<A> acquire,
    Func<A, IO<B>> release,
    Func<A, IO<C>> action) =>
    IO.lift(env =>
    {
        var x = acquire.Run(env);
        try
        {
            return action(x).Run(env);
        }
        finally
        {
            release(x).Run(env);
        }
    });
```
If the acquired resource is `IDisposable` then you can write a `bracket` function that doesn't need a `release` lambda argument:
```c#
public static IO<B> bracket<A, B>(
    IO<A> acquire,
    Func<A, IO<B>> action)
    where A : IDisposable =>
    IO.lift(env =>
    {
        var x = acquire.Run(env);
        try
        {
            return action(x).Run(env);
        }
        finally
        {
            x.Dispose();
        }
    });
```

This is using the IO monad from `v5` (which is in alpha right now).  But, the same principle will work for any monad in v4 too.

Support for a `bracket` function is in my [v5 TODO-list](https://github.com/louthy/language-ext/tree/v5-transducers/Major%20Version%20Release%20Notes/Version%205#todo).  `v5` also has a [`ResourceT` monad transformer](https://github.com/louthy/language-ext/tree/v5-transducers/LanguageExt.Core/Traits/Resource) that automates tracking and releasing of resources - you can see the functions it supports [here](https://github.com/louthy/language-ext/blob/v5-transducers/LanguageExt.Core/Traits/Resource/Resource.Module.cs).  If you [look at the implementation of `Eff<RT, A>`](https://github.com/louthy/language-ext/blob/v5-transducers/LanguageExt.Core/Effects/Eff/Eff%20with%20runtime/Eff.cs#L17) you'll see how its implemented as a monad-transformer stack: `StateT<ResourceT<IO, A>>` -- which means `Eff` does resource-tracking automatically.

NOTE: None of this is tested yet.  And if there's one thing I expect to not be perfect in the alpha it's the resource-tracking, so be aware that it's likely to have some bugs.

In `v4` the resource tracking was almost non-existent.  Pipes did auto-resource tracking, but nothing else.  All other monadic types (that were computational and not pure data) had their own `use` functions.  [Here's the `Eff` and `Aff` `use` functions](https://github.com/louthy/language-ext/blob/main/LanguageExt.Core/Effects/Eff/Prelude/Eff.Prelude.use.cs) -- you'll see they're implemented like the `bracket` function above.

Resource tracking is one of the _'big pushes'_ of the `v5` release; now that we can do higher-kinds and monad-transformer stacking, we can put resource management in the hands of the dev building their stack, rather than it all being prescribed by the monad implementations in language-ext.

You didn't give a huge amount of context, but hopefully that answers the question!
